lib/Mir.ml

```diff
@@ -13,76 +13,59 @@ open Core_kernel
 
 type litType = Int | Real | Str [@@deriving sexp, hash]
 
-type operator =
-  | Plus
-  | Minus
-  | Times
-  | Divide
-  | Modulo
-  | Or
-  | And
-  | Equals
-  | NEquals
-  | Less
-  | Leq
-  | Greater
-  | Geq
-[@@deriving sexp, hash]
-
-type transformation = expr Ast.transformation
-
-and index =
+and 'e index =
   | All
-  | Single of expr
+  | Single of 'e
   (*
-  | MatrixSingle of expr
+  | MatrixSingle of 'e
  *)
-  | Upfrom of expr
-  | Downfrom of expr
-  | Between of expr * expr
-  | MultiIndex of expr
+  | Upfrom of 'e
+  | Downfrom of 'e
+  | Between of 'e * 'e
+  | MultiIndex of 'e
 
 (** XXX
 *)
-and expr =
+and 'e expr =
   | Var of string
   | Lit of litType * string
-  | FunApp of string * expr list
-  | BinOp of expr * operator * expr
-  | TernaryIf of expr * expr * expr
-  | Indexed of expr * index list
+  | FunApp of string * 'e list
+  | TernaryIf of 'e * 'e * 'e
+  | Indexed of 'e * 'e index list
 [@@deriving sexp, hash]
 
-type adtype = Ast.autodifftype [@@deriving sexp, hash]
-type sizedtype = expr Ast.sizedtype [@@deriving sexp, hash]
 type unsizedtype = Ast.unsizedtype [@@deriving sexp, hash]
+type autodifftype = Ast.autodifftype [@@deriving sexp, hash]
+
+let no_loc = {Ast.filename= ""; line_num= 0; col_num= 0; included_from= None}
+let no_span = {Ast.begin_loc= no_loc; end_loc= no_loc}
 
 (* This directive silences some spurious warnings from ppx_deriving *)
 [@@@ocaml.warning "-A"]
 
-type fun_arg_decl = (adtype * string * unsizedtype) list
+type fun_arg_decl = (autodifftype * string * unsizedtype) list
 
-and 's statement =
-  | Assignment of expr * expr
-  | TargetPE of expr
-  | NRFunApp of string * expr list
-  | Check of string * expr list
+and ('e, 's) statement =
+  | Assignment of 'e * 'e
+  | TargetPE of 'e
+  | NRFunApp of string * 'e list
+  | Check of string * 'e list
   | Break
   | Continue
-  | Return of expr option
+  | Return of 'e option
   | Skip
-  | IfElse of expr * 's * 's option
-  | While of expr * 's
+  | IfElse of 'e * 's * 's option
+  | While of 'e * 's
   (* XXX Collapse with For? *)
-  | For of {loopvar: expr; lower: expr; upper: expr; body: 's}
+  | For of {loopvar: 'e; lower: 'e; upper: 'e; body: 's}
```

Comment:
I still don't understand why the loopvar is an expression here, rather than a plain string.

Reply:
I'm forgetting the previous conversation, sorry! I can't think of a reason a string wouldn't work right now. Anyone else?

Reply:
No prob! Not 100% sure I'd asked you about it before. Maybe I just wondered about it in my head.